build fails on ubuntu

Please see below log

```
> opendds@0.2.1 install
> node-gyp rebuild

gyp info it worked if it ends with ok
gyp info using node-gyp@9.1.0
gyp info using node@16.18.1 | linux | x64
gyp info find Python using Python version 3.10.6 found at "/usr/bin/python3"
gyp info spawn /usr/bin/python3
gyp info spawn args [
gyp info spawn args   '/usr/lib/node_modules/npm/node_modules/node-gyp/gyp/gyp_main.py',
gyp info spawn args   'binding.gyp',
gyp info spawn args   '-f',
gyp info spawn args   'make',
gyp info spawn args   '-I',
gyp info spawn args   '/home/ubuntu/source/node-opendds/build/config.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '/usr/lib/node_modules/npm/node_modules/node-gyp/addon.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '/home/ubuntu/.cache/node-gyp/16.18.1/include/node/common.gypi',
gyp info spawn args   '-Dlibrary=shared_library',
gyp info spawn args   '-Dvisibility=default',
gyp info spawn args   '-Dnode_root_dir=/home/ubuntu/.cache/node-gyp/16.18.1',
gyp info spawn args   '-Dnode_gyp_dir=/usr/lib/node_modules/npm/node_modules/node-gyp',
gyp info spawn args   '-Dnode_lib_file=/home/ubuntu/.cache/node-gyp/16.18.1/<(target_arch)/node.lib',
gyp info spawn args   '-Dmodule_root_dir=/home/ubuntu/source/node-opendds',
gyp info spawn args   '-Dnode_engine=v8',
gyp info spawn args   '--depth=.',
gyp info spawn args   '--no-parallel',
gyp info spawn args   '--generator-output',
gyp info spawn args   'build',
gyp info spawn args   '-Goutput_dir=.'
gyp info spawn args ]
gyp info spawn make
gyp info spawn args [ 'BUILDTYPE=Release', '-C', 'build' ]
make: Entering directory '/home/ubuntu/source/node-opendds/build'
  CXX(target) Release/obj.target/node_opendds/src/node-opendds.o
In file included from /home/ubuntu/source/OpenDDS-3.22/dds/DCPS/DataReaderImpl.h:20,
                 from ../src/NodeDRListener.h:11,
                 from ../src/node-opendds.cpp:1:
/home/ubuntu/source/OpenDDS-3.22/dds/DCPS/DomainParticipantImpl.h:34:12: fatal error: dds/DdsDcpsCoreTypeSupportC.h: No such file or directory
   34 | #  include <dds/DdsDcpsCoreTypeSupportC.h>
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make: *** [node_opendds.target.mk:130: Release/obj.target/node_opendds/src/node-opendds.o] Error 1
make: Leaving directory '/home/ubuntu/source/node-opendds/build'
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:201:23)
gyp ERR! stack     at ChildProcess.emit (node:events:513:28)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (node:internal/child_process:293:12)
gyp ERR! System Linux 5.15.0-1019-aws
gyp ERR! command "/usr/bin/node" "/usr/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /home/ubuntu/source/node-opendds
gyp ERR! node -v v16.18.1
gyp ERR! node-gyp -v v9.1.0
gyp ERR! not ok 
npm ERR! code 1
npm ERR! path /home/ubuntu/source/node-opendds
npm ERR! command failed
npm ERR! command sh -c -- node-gyp rebuild

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/ubuntu/.npm/_logs/...-debug-0.log
```

I used `setenv.sh` in the OpenDDS-3.22 folder

Ok, so I tried rebuilding OpenDDS3.22 with configuration

```
./configure --built-in-topics --no-tests --no-debug --no-rapidjson --no-content-subscription --no-content-filtered-topic --no-multi-topic --no-query-condition --no-ownership-profile --no-ownership-kind-exclusive --no-object-model-profile --no-persistence-profile 
source ./setenv.sh
make -j $(getconf _NPROCESSORS_ONLN) INSTALL_PREFIX=/usr/local 
```

as per your suggestion, but now I am stuck with linker errors in building opendds...

```
/usr/bin/ld: /home/ubuntu/source/OpenDDS-3.22/ACE_wrappers/lib/libTAO_IDL_FE.so: undefined reference to `__cxa_throw_bad_array_new_length'
/usr/bin/ld: /home/ubuntu/source/OpenDDS-3.22/ACE_wrappers/lib/libTAO_IDL_FE.so: undefined reference to `std::basic_ios<char, std::char_traits<char> >::rdbuf(std::basic_streambuf<char, std::char_traits<char> >*)'
/usr/bin/ld: /home/ubuntu/source/OpenDDS-3.22/lib/libOpenDDS_Util.so: undefined reference to `std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >::c_str() const'
/usr/bin/ld: warning: creating DT_TEXTREL in a PIE
collect2: error: ld returned 1 exit status
make[1]: *** [GNUmakefile.opendds_idl:225: opendds_idl] Error 1
make[1]: Leaving directory '/home/ubuntu/source/OpenDDS-3.22/dds/idl'
make: *** [GNUmakefile.dist:471: opendds_idl] Error 2
```

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -22,6 +22,8 @@ jobs:
         node_version:
           - 18
           - 20
+          - 22
+          - 24
         opendds_branch:
           - master
           - latest-release

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -126,7 +126,7 @@ jobs:
             CONFIG_OPTIONS+=" --security"
             BUILD_TARGETS+=" OpenDDS_Security"
             if [ '${{ matrix.m.os }}' == 'macos-14' ]; then
-              CONFIG_OPTIONS+=" --xerces3=$(brew --prefix xerces-c) --openssl=/usr/local/opt/openssl@1.1"
+              CONFIG_OPTIONS+=" --xerces3=$(brew --prefix xerces-c) --openssl=$(brew --prefix openssl@3)"
             fi
           fi
           echo "CONFIG_OPTIONS=$CONFIG_OPTIONS" >> $GITHUB_ENV

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -83,7 +83,7 @@ jobs:
         uses: lukka/run-vcpkg@v11
         with:
           vcpkgDirectory: '${{ github.workspace }}/vcpkg'
-          vcpkgGitCommitId: 898b728edc5e0d12b50015f9cd18247c4257a3eb
+          vcpkgGitCommitId: 940f58770cee8e2011bfb4847fb2bd70057301a8
           runVcpkgInstall: true
       - name: 'Set Up MSVC Environment'
         if: ${{ matrix.m.os == 'windows-2022' }}

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -30,8 +30,8 @@ jobs:
         m:
           - {os: ubuntu-22.04, dds_security: 1}
           - {os: ubuntu-22.04, dds_security: 0}
-          - {os: macos-13, dds_security: 1}
-          - {os: macos-13, dds_security: 0}
+          - {os: macos-14, dds_security: 1}
+          - {os: macos-14, dds_security: 0}
           - {os: windows-2022, dds_security: 1}
           - {os: windows-2022, dds_security: 0}
 
@@ -85,7 +85,7 @@ jobs:
           sudo apt-get update
           sudo apt-get -y install libxerces-c-dev
       - name: 'Install xerces (macos)'
-        if: ${{ matrix.m.dds_security == 1 && matrix.m.os == 'macos-13' }}
+        if: ${{ matrix.m.dds_security == 1 && matrix.m.os == 'macos-14' }}
         run: |
           brew install xerces-c
       - name: Setup for run-vcpkg
@@ -109,10 +109,10 @@ jobs:
         if: ${{ matrix.m.os == 'windows-2022' }}
         uses: ammaraskar/msvc-problem-matcher@0.3.0
       - name: 'Set Up Problem Matcher (ubuntu / macos)'
-        if: ${{ matrix.m.os == 'ubuntu-22.04' || matrix.m.os == 'macos-13' }}
+        if: ${{ matrix.m.os == 'ubuntu-22.04' || matrix.m.os == 'macos-14' }}
         uses: ammaraskar/gcc-problem-matcher@0.3.0
       - name: 'Set environment variables (ubuntu / macos)'
-        if: ${{ matrix.m.os == 'ubuntu-22.04' || matrix.m.os == 'macos-13' }}
+        if: ${{ matrix.m.os == 'ubuntu-22.04' || matrix.m.os == 'macos-14' }}
         shell: bash
         run: |-
           echo "ACE_ROOT=$GITHUB_WORKSPACE/ACE_TAO/ACE" >> $GITHUB_ENV
@@ -125,7 +125,7 @@ jobs:
           if [ ${{ matrix.m.dds_security }} == 1 ]; then
             CONFIG_OPTIONS+=" --security"
             BUILD_TARGETS+=" OpenDDS_Security"
-            if [ '${{ matrix.m.os }}' == 'macos-13' ]; then
+            if [ '${{ matrix.m.os }}' == 'macos-14' ]; then
               CONFIG_OPTIONS+=" --xerces3=$(brew --prefix xerces-c) --openssl=/usr/local/opt/openssl@1.1"
             fi
           fi
@@ -149,7 +149,7 @@ jobs:
           echo "BUILD_TARGETS=DCPSInfoRepo_Main;OpenDDS_Rtps_Udp$BUILD_TARGETS" >> $GITHUB_ENV
           echo "ACE_TEST_LOG_STUCK_STACKS=1" >> $GITHUB_ENV
       - name: 'Configure & Build OpenDDS (ubuntu / macos)'
-        if: ${{ (matrix.m.os == 'ubuntu-22.04' || matrix.m.os == 'macos-13') && steps.cache-opendds.outputs.cache-hit != 'true' }}
+        if: ${{ (matrix.m.os == 'ubuntu-22.04' || matrix.m.os == 'macos-14') && steps.cache-opendds.outputs.cache-hit != 'true' }}
         shell: bash
         run: |-
           echo "dds_security=${{ matrix.m.dds_security }}; CONFIG_OPTIONS=${{ env.CONFIG_OPTIONS }}; BUILD_TARGETS=${{ env.BUILD_TARGETS }}"
@@ -174,7 +174,7 @@ jobs:
           call setenv.cmd
           msbuild -p:Configuration=Debug,Platform=x64 -t:${{ env.BUILD_TARGETS }} -m DDS_TAOv2.sln
       - name: 'Install (ubuntu / macos)'
-        if: ${{ matrix.m.os == 'ubuntu-22.04' || matrix.m.os == 'macos-13' }}
+        if: ${{ matrix.m.os == 'ubuntu-22.04' || matrix.m.os == 'macos-14' }}
         shell: bash
         run: |-
           npm install
@@ -195,12 +195,12 @@ jobs:
           echo Core file pattern set to:
           cat /proc/sys/kernel/core_pattern
       - name: 'Change Cores Directory Permissions (macos)'
-        if: ${{ matrix.m.os == 'macos-13' }}
+        if: ${{ matrix.m.os == 'macos-14' }}
         shell: bash
         run: |
           sudo chmod o+w /cores
       - name: 'Test (ubuntu / macos)'
-        if: ${{ matrix.m.os == 'ubuntu-22.04' || matrix.m.os == 'macos-13' }}
+        if: ${{ matrix.m.os == 'ubuntu-22.04' || matrix.m.os == 'macos-14' }}
         shell: bash
         run: |-
           ulimit -c unlimited
@@ -212,7 +212,7 @@ jobs:
           ./run_test.pl node2cpp --rtps
           ./run_test.pl node2node --rtps
       - name: 'Test Secure (ubuntu / macos)'
-        if: ${{ (matrix.m.os == 'ubuntu-22.04' || matrix.m.os == 'macos-13') && matrix.m.dds_security == 1 }}
+        if: ${{ (matrix.m.os == 'ubuntu-22.04' || matrix.m.os == 'macos-14') && matrix.m.dds_security == 1 }}
         shell: bash
         run: |-
           ulimit -c unlimited

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -62,6 +62,23 @@ jobs:
         uses: actions/setup-node@v4
         with:
           node-version: ${{ matrix.node_version }}
+          cache: 'npm'
+      - name: 'Get Repository SHAs'
+        id: get-shas
+        shell: bash
+        run: |
+          echo "opendds_sha=$(cd OpenDDS && git rev-parse HEAD)" >> $GITHUB_OUTPUT
+          echo "acetao_sha=$(cd ACE_TAO && git rev-parse HEAD)" >> $GITHUB_OUTPUT
+          echo "mpc_sha=$(cd MPC && git rev-parse HEAD)" >> $GITHUB_OUTPUT
+      - name: 'Cache OpenDDS Build'
+        uses: actions/cache@v4
+        id: cache-opendds
+        with:
+          path: |
+            ACE_TAO
+            OpenDDS
+            MPC
+          key: opendds-build-${{ matrix.m.os }}-${{ matrix.opendds_branch }}-sec${{ matrix.m.dds_security }}-${{ steps.get-shas.outputs.opendds_sha }}-${{ steps.get-shas.outputs.acetao_sha }}
       - name: 'Install xerces (ubuntu)'
         if: ${{ matrix.m.dds_security == 1 && matrix.m.os == 'ubuntu-22.04' }}
         run: |-
@@ -132,7 +149,7 @@ jobs:
           echo "BUILD_TARGETS=DCPSInfoRepo_Main;OpenDDS_Rtps_Udp$BUILD_TARGETS" >> $GITHUB_ENV
           echo "ACE_TEST_LOG_STUCK_STACKS=1" >> $GITHUB_ENV
       - name: 'Configure & Build OpenDDS (ubuntu / macos)'
-        if: ${{ matrix.m.os == 'ubuntu-22.04' || matrix.m.os == 'macos-13' }}
+        if: ${{ (matrix.m.os == 'ubuntu-22.04' || matrix.m.os == 'macos-13') && steps.cache-opendds.outputs.cache-hit != 'true' }}
         shell: bash
         run: |-
           echo "dds_security=${{ matrix.m.dds_security }}; CONFIG_OPTIONS=${{ env.CONFIG_OPTIONS }}; BUILD_TARGETS=${{ env.BUILD_TARGETS }}"
@@ -142,7 +159,7 @@ jobs:
           . setenv.sh
           make -j$(getconf _NPROCESSORS_ONLN) ${{ env.BUILD_TARGETS }}
       - name: 'Configure OpenDDS (windows)'
-        if: ${{ matrix.m.os == 'windows-2022' }}
+        if: ${{ matrix.m.os == 'windows-2022' && steps.cache-opendds.outputs.cache-hit != 'true' }}
         shell: cmd
         run: |-
           echo "dds_security=${{ matrix.m.dds_security }}; CONFIG_OPTIONS=${{ env.CONFIG_OPTIONS }}; BUILD_TARGETS=${{ env.BUILD_TARGETS }}"
@@ -150,7 +167,7 @@ jobs:
           configure --no-tests ${{ env.CONFIG_OPTIONS }}
           perl tools/scripts/show_build_config.pl
       - name: 'Build OpenDDS (windows)'
-        if: ${{ matrix.m.os == 'windows-2022' }}
+        if: ${{ matrix.m.os == 'windows-2022' && steps.cache-opendds.outputs.cache-hit != 'true' }}
         shell: cmd
         run: |-
           cd OpenDDS

--- a/README.md
+++ b/README.md
@@ -8,7 +8,7 @@
     * If using security, you may need to ensure that OpenDDS is built with the same version of OpenSSL used by Node.js
 
 ## Tested platforms:
-* Node LTS versions 8, 10, 12, 14, 16, 18, 20, 22, 24
+* Node LTS versions 18, 20, 22, 24
 * Linux (Ubuntu 20.04) x86_64 using gcc 9.4.0
 * Linux (Ubuntu 22.04) x86_64 using gcc 11.2.0 (w/ openssl-1.1.1q)
 * Windows (Server 2022) x86_64 using Visual Studio Enterprise 2022

--- a/README.md
+++ b/README.md
@@ -8,12 +8,24 @@
     * If using security, you may need to ensure that OpenDDS is built with the same version of OpenSSL used by Node.js
 
 ## Tested platforms:
-* Node LTS versions 8, 10, 12, 14, 16, 18, 20
+* Node LTS versions 8, 10, 12, 14, 16, 18, 20, 22, 24
 * Linux (Ubuntu 20.04) x86_64 using gcc 9.4.0
 * Linux (Ubuntu 22.04) x86_64 using gcc 11.2.0 (w/ openssl-1.1.1q)
 * Windows (Server 2022) x86_64 using Visual Studio Enterprise 2022
 * macOS (11.7) x86_54 using clang 13.0
 * Other OpenDDS-supported platforms should work, but may required changes to binding.gyp
+
+## Quality of Service (QoS)
+When passing QoS objects to entities, note that fields defined as sequences (like `representation`) must be passed as arrays, even if they only contain a single value. For example, to set the `representation` QoS for a DataWriter, the property should be structured like this:
+
+```javascript
+let qos = {
+  // ... other QoS settings
+  representation: {
+    value: ["DDS::XCDR_DATA_REPRESENTATION"]
+  }
+};
+```
 
 ## Building and running the tests
 * In OpenDDS directory:

--- a/README.md
+++ b/README.md
@@ -52,6 +52,12 @@ $ node-gyp configure build
 
 ## Changelog
 
+### Version 0.3.1
+
+* Added support for Node.js 22 and 24 (Updated `nan` dependency)
+* Fixed an issue where invalid-data callbacks included misleading sample payloads
+* Implemented map types in `ValueReader` and `ValueWriter` to satisfy updated OpenDDS interfaces
+
 ### Version 0.3.0
 
 * Update NodeValueWriter and NodeValueReader accordingly to the ValueWriter and ValueReader changes in OpenDDS 3.28.0

--- a/package.json
+++ b/package.json
@@ -28,6 +28,6 @@
   "license": "BSD-3-Clause",
   "gypfile": true,
   "dependencies": {
-    "nan": "^2.19.0"
+    "nan": "^2.26.2"
   }
 }

--- a/src/NodeValueReader.cpp
+++ b/src/NodeValueReader.cpp
@@ -191,6 +191,36 @@ bool NodeValueReader::end_element()
   return true;
 }
 
+bool NodeValueReader::begin_map(OpenDDS::XTypes::TypeKind /*key_kind*/, OpenDDS::XTypes::TypeKind /*value_kind*/)
+{
+  return false;
+}
+
+bool NodeValueReader::end_map()
+{
+  return false;
+}
+
+bool NodeValueReader::begin_key()
+{
+  return false;
+}
+
+bool NodeValueReader::end_key()
+{
+  return false;
+}
+
+bool NodeValueReader::begin_value()
+{
+  return false;
+}
+
+bool NodeValueReader::end_value()
+{
+  return false;
+}
+
 bool NodeValueReader::read_boolean(ACE_CDR::Boolean& value)
 {
   return primitive_helper<v8::Boolean>(value, &v8::Value::IsBoolean);

--- a/src/NodeValueReader.h
+++ b/src/NodeValueReader.h
@@ -49,6 +49,13 @@ public:
   bool begin_element();
   bool end_element();
 
+  bool begin_map(OpenDDS::XTypes::TypeKind key_kind, OpenDDS::XTypes::TypeKind value_kind);
+  bool end_map();
+  bool begin_key();
+  bool end_key();
+  bool begin_value();
+  bool end_value();
+
   bool read_boolean(ACE_CDR::Boolean& value);
   bool read_byte(ACE_CDR::Octet& value);
 #if OPENDDS_HAS_EXPLICIT_INTS

--- a/src/NodeValueWriter.cpp
+++ b/src/NodeValueWriter.cpp
@@ -121,6 +121,36 @@ bool NodeValueWriter::end_element()
   return true;
 }
 
+bool NodeValueWriter::begin_map(OpenDDS::XTypes::TypeKind /*key_kind*/, OpenDDS::XTypes::TypeKind /*value_kind*/)
+{
+  return false;
+}
+
+bool NodeValueWriter::end_map()
+{
+  return false;
+}
+
+bool NodeValueWriter::begin_key()
+{
+  return false;
+}
+
+bool NodeValueWriter::end_key()
+{
+  return false;
+}
+
+bool NodeValueWriter::begin_value()
+{
+  return false;
+}
+
+bool NodeValueWriter::end_value()
+{
+  return false;
+}
+
 bool NodeValueWriter::write_boolean(ACE_CDR::Boolean value)
 {
   return primitive_helper<ACE_CDR::Boolean, v8::Boolean>(value);

--- a/src/NodeValueWriter.h
+++ b/src/NodeValueWriter.h
@@ -36,6 +36,13 @@ public:
   bool begin_element(ACE_CDR::ULong idx);
   bool end_element();
 
+  bool begin_map(OpenDDS::XTypes::TypeKind key_kind, OpenDDS::XTypes::TypeKind value_kind);
+  bool end_map();
+  bool begin_key();
+  bool end_key();
+  bool begin_value();
+  bool end_value();
+
   bool write_boolean(ACE_CDR::Boolean value);
   bool write_byte(ACE_CDR::Octet value);
 #if OPENDDS_HAS_EXPLICIT_INTS

